lib/core/CProgramCounters.cc

```diff
@@ -0,0 +1,192 @@
+
+#include <rapidjson/document.h>
+#include <rapidjson/ostreamwrapper.h>
+
+#include <ostream>
+#include <string>
+
+namespace ml {
+namespace core {
+
+namespace {
+
+using TGenericLineWriter = core::CRapidJsonLineWriter<rapidjson::OStreamWrapper>;
+
+const std::string NAME_TYPE("name");
+const std::string DESCRIPTION_TYPE("description");
+const std::string COUNTER_TYPE("value");
+
+//! Persistence tags
+const std::string KEY_TAG("a");
+const std::string VALUE_TAG("b");
+
+//! Helper function to add a string/int pair to JSON writer
+void addStringInt(TGenericLineWriter& writer,
+                  const std::string& name,
+                  const std::string& description,
+                  uint64_t counter) {
+    writer.StartObject();
+
+    writer.String(NAME_TYPE);
+    writer.String(name);
+
+    writer.String(DESCRIPTION_TYPE);
+    writer.String(description);
+
+    writer.String(COUNTER_TYPE);
+    writer.Uint64(counter);
+
+    writer.EndObject();
+}
+}
+
+CProgramCounters::CProgramCounters()
+    : m_Counters(), m_Cache(), m_ProgramCounterTypes() {
+}
+
+CProgramCounters& CProgramCounters::instance() {
+    return ms_Instance;
+}
+
+CProgramCounters::TCounter& CProgramCounters::counter(counter_t::ECounterTypes counterType) {
+    if (counterType >= counter_t::E_LastEnumCounter) {
+        LOG_ABORT(<< "Bad index " << counterType);
+    }
+    return ms_Instance.m_Counters[counterType];
+}
+
+CProgramCounters::TCounter& CProgramCounters::counter(size_t index) {
+    if (index >= ms_Instance.m_Counters.size()) {
+        LOG_ABORT(<< "Bad index " << index);
+    }
+    return ms_Instance.m_Counters[index];
+}
+
+void CProgramCounters::cacheCounters() {
+    if (ms_Instance.m_Cache.size() != 0) {
+        // The cache should only exist for a very brief period of time,
+        // while persistence is in operation.
+        // Elsewhere there are checks to ensure that only one persistence
+        // operation occurs at any point in time,so the cache _should_ be
+        // empty at this point.  Warn if that isn't the case.
+        LOG_WARN(<< "Overwriting existing counters cache.");
+
+        // clear the cache
+        TUInt64Vec().swap(ms_Instance.m_Cache);
+    }
+    ms_Instance.m_Cache.assign(ms_Instance.m_Counters.begin(),
+                               ms_Instance.m_Counters.end());
+}
+
+void CProgramCounters::staticsAcceptPersistInserter(CStatePersistInserter& inserter) {
+    // As the analytics thread could be updating counters while this method is running, in order
+    // to guarantee that consistent counters get persisted for a background persistence task we must not persist
+    // live values. Instead, we access a cached set of self consistent counters for persistence.
+
+    auto staticsAcceptPersistInserter = [&](const auto& container) {
+        int i = 0;
+        for (const auto& counter : container) {
+            inserter.insertValue(KEY_TAG, i);
+            inserter.insertValue(VALUE_TAG, counter);
```

Comment:
if the counter is `0` - could we skip it?

Reply:
Also, perhaps only write out counter the program has registered?

Reply:
One thing to confirm before doing this is what the affect on BWC would be.  Would 6.7 tolerate restoring an incomplete set of counters?